Oxb 199/glob pathing

rpschoenburg opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Improvements**
  * Unified glob/path expansion across add, rm, restore, and remote workflows for consistent wildcard behavior and parity between local and remote modes.
  * Improved relative-path validation and handling when operating inside vs outside a repository.

* **New Features**
  * Public GlobOpts configuration and reusable glob-expansion API for flexible wildcard resolution.
  * Default constructor for removal options to simplify CLI/default usage.

* **Tests**
  * Added tests for wildcard add/rm/restore scenarios and staged-file behaviors.

* **Chores**
  * Minor logging, formatting, and whitespace cleanups.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rpschoenburg committed:
Initial refactor of globbing logic

rpschoenburg committed:
Implement glob_opts for glob path parsing

Introduces the GlobOpts struct, which allows functions to control where
glob paths are matched against. Also includes corrections for unit
tests, including having non-glob dirs be walked for paths if walk_dirs
is set, which is necessary for workspace add. As well, corrections for
different current working directories are included

rpschoenburg committed:
Merge branch 'main' into OXB-199/glob_pathing

rpschoenburg committed:
Linters

rpschoenburg committed:
Correct glob path parsing module for unit tests

rpschoenburg committed:
Merge branch 'main' into OXB-199/glob_pathing

rpschoenburg committed:
Linters

jcelliott committed:
Merge branch 'main' into OXB-199/glob_pathing

jcelliott commented:
@rpschoenburg it looks like `repositories::remote_mode::commit::tests::test_remote_mode_commit_removed_dir` is failing consistently, so probably an issue with the glob pathing. Can you take a look?

Lilianxr commented on oxen-rust/src/lib/src/util/glob.rs:
```diff
@@ -0,0 +1,363 @@
+            return Ok(());
+        }
+
+        // Else, recurse into the matching directories
+        let mut new_index = *search_index + 1;
+        for mut entry in matched_entries {
+            r_search_merkle_tree(
+                repo,
+                head_commit,
+                glob_path_components,
+                paths,
+                &mut entry,
+                &mut new_index,
+            )?;
```
@rpschoenburg Hi Ryan, I think this is a good catch

Lilianxr commented on oxen-rust/src/lib/src/util/fs.rs:
```diff
@@ -1585,6 +1521,67 @@ pub fn path_relative_to_dir(
+
+    let mut dir_component = dir_iter.next().unwrap();
+    let mut matches = 0;
+
+    for _ in 0..(path_components.len()) {
+        let path_component = path_iter.next().expect("Path bounds violated");
+        let path_str = path_component.as_os_str();
+        let dir_str = dir_component.as_os_str();
+
+        if path_str == dir_str {
+            matches += 1;
+            if matches == dir_components.len() {
+                return true;
+            }
+            dir_component = dir_iter.next().expect("Dir bounds violated");
+            continue;
+        }
+
+        if path_str.len() == dir_str.len() {
+            let path_lower = path_str.to_string_lossy().to_lowercase();
+            let dir_lower = dir_str.to_string_lossy().to_lowercase();
+
+            if path_lower == dir_lower {
+                matches += 1;
+                if matches == dir_components.len() {
+                    return true;
+                }
+                dir_component = dir_iter.next().expect("Dir bounds violated");
+                continue;
+            }
+        }
+
+        // If the components don't match, reset dir_iter and dir_component
+        dir_iter = starting_dir_iter.clone();
+        dir_component = dir_iter.next().unwrap();
+    }
+
+    // If the loop finishes, the path cannot be found relative to the dir
+    false
+}
```
This is worth noting

jcelliott commented on oxen-rust/src/lib/src/opts/glob_opts.rs:
```diff
@@ -0,0 +1,10 @@
+use std::path::PathBuf;
+
+#[derive(Clone, Debug)]
+pub struct GlobOpts {
```
@rpschoenburg can you document what these options do here? It's not obvious to me what they all mean.

We should also have `Default` trait implementation for the struct (can just be derived) and then everywhere we are manually building a `GlobOpts` we should use the default implementation like this:
```
let glob_opts = GlobOpts {
    paths: my_paths,
    ..Default::default()
};
```
That makes it so changes in the GlobOpts struct in the future don't break the existing functionality.

jcelliott commented:
@rpschoenburg comparing the output from the CI run vs my local, I noticed a difference that likely points to the problem:

**Local** 
```
cargo test --lib repositories::remote_mode::commit::tests::test_remote_mode_commit_removed_dir -- --nocapture
```
```
...
Remote-mode repository initialized successfully!
Workspace ID: bafa0f9e-78dc-44fd-a825-5147708ab10e
Fetch remote branch: repo_503f8752-5b70-4cfa-8cb9-b70a8f9fffa1/main
Files to send: 2
🐂 oxen added 2 entries to workspace main: bafa0f9e-78dc-44fd-a825-5147708ab10e
Committing to remote with message: Adding train directory
...
```

**CI**
```
...
Remote-mode repository initialized successfully!
Workspace ID: 8530ccb0-adc0-44f5-b3e1-4387019deadf
Fetch remote branch: repo_a45b118b-85f7-4c47-a5a9-b7fcb03c5407/main
🐂 oxen added 1 entries to workspace main: 8530ccb0-adc0-44f5-b3e1-4387019deadf
Committing to remote with message: Adding train directory
Error running test. Err: Basic(Basic(No changes to commit))
```

We're not getting the `Files to send: 2` message in CI. Looking at the code, that probably means that the files failed to upload for some reason or possibly were filtered out by the glob logic. I would put some debugging print statements in there and rerun in CI to see what you get. You could also temporarily change the test job to only run this one test to make debugging easier.

rpschoenburg committed:
Fixes for CodeRabbit

rpschoenburg committed:
Fix handling of oxen add .

rpschoenburg committed:
Fix for unit test

rpschoenburg committed:
Add unit tests for dot paths

rpschoenburg committed:
Add extra unit tests for wildcard paths

rpschoenburg committed:
Add unit tests for glob module

rpschoenburg committed:
Merge branch 'main' into OXB-199/glob_pathing

rpschoenburg commented:
To test this PR, I'd recommend doing add, rm, restore, or workspace/remote-mode add, rm, and restore with wildcard paths. The idea of this PR is that anything that you could use with git should also work for oxen.

With regards to the specific issue that inspired this PR (Adding removed paths with '.'), if you do an add that matches against both new files and removed files, the print message with the stats for each will show up. I'm not sure that's ideal, and I considered having the print statement for rm not show up, but for now it's still there.

As well, `oxen restore .` won't work, because oxen restore doesn't currently use the same path expansion at the cmd level as add and rm do. That could easily be changed.

At one point, I accidentally committed a file from my oxen mv PR. I removed it again in this PR, but it's in the commit history

rpschoenburg committed:
Linters

Lilianxr commented on oxen-rust/src/lib/src/util/glob.rs:
```diff
@@ -0,0 +1,626 @@
+        // Paths ending in '.' are expanded to the current dir at the cmd level
+        log::debug!("path: {path:?}");
+        let path = if *path == repo_path {
+            path.join("*")
```
'.' is not a glob pattern right, why do we handle it differently here?

rpschoenburg committed:
Refactor logic for adding removed paths with .

rpschoenburg committed:
Recursively check for removed files in collect_removed_paths

rpschoenburg committed:
Merge branch 'main' into OXB-199/glob_pathing
